README.md
# PhonyRails [![Build Status](https://travis-ci.org/joost/phony_rails.svg?branch=master)](https://travis-ci.org/joost/phony_rails) [![Coverage Status](https://coveralls.io/repos/joost/phony_rails/badge.svg)](https://coveralls.io/r/joost/phony_rails) ![Dependencies Status](https://img.shields.io/gem/v/phony_rails.svg)

This small Gem adds useful methods to your Rails app to validate, display and save phone numbers.
It uses the super awesome Phony gem (https://github.com/floere/phony).

Find version information in the [CHANGELOG](CHANGELOG.md).

## Installation

Add this line to your application's Gemfile (requires Ruby > 2.3):

```ruby
gem 'phony_rails'
```

And then execute:

```
$ bundle
```

Or install it yourself as:

```
$ gem install phony_rails
```

## Usage

### Normalization / Model Usage

#### ActiveRecord

For **ActiveRecord**, in your model add:

```ruby
class SomeModel < ActiveRecord::Base
  # Normalizes the attribute itself before validation
  phony_normalize :phone_number, default_country_code: 'US'

  # Normalizes attribute before validation and saves into other attribute
  phony_normalize :phone_number, as: :phone_number_normalized_version, default_country_code: 'US'

  # Creates method normalized_fax_number that returns the normalized version of fax_number
  phony_normalized_method :fax_number

  # Conditionally normalizes the attribute
  phony_normalize :recipient, default_country_code: 'US', if: -> { contact_method == 'phone_number' }
end
```

#### ActiveModel (models without database)

For Rails-like models without a database, add:

```ruby
class SomeModel
  include ActiveModel::Model # we get AR-like attributes and validations
  include ActiveModel::Validations::Callbacks # a dependency for normalization

  # your attributes must be defined, they are not inherited from a DB table
  attr_accessor :phone_number, :phone_number_as_normalized

  # Once the model is set up, we have the same things as with ActiveRecord
  phony_normalize :phone_number, default_country_code: 'US'
end
```

#### Mongoid (DEPRECATED)

WARNING: From v0.15.0 Mongoid support has been removed!

#### General info

The `:default_country_code` options is used to specify a country_code when normalizing.

PhonyRails will also check your model for a country_code method to use when normalizing the number. So `'070-12341234'` with `country_code` 'NL' will get normalized to `'+317012341234'`.

You can also do-it-yourself and call:

```ruby
# Options:
#   :country_code => The country code we should use (forced).
#   :default_country_code => Some fallback code (eg. 'NL') that can be used as default (comes from phony_normalize_numbers method).

PhonyRails.normalize_number('some number', country_code: 'NL')

PhonyRails.normalize_number('+4790909090', country_code: 'SE') # => '+464790909090' (forced to +46)
PhonyRails.normalize_number('+4790909090', default_country_code: 'SE') # => '+4790909090' (still +47 so not changed)
```

The country_code should always be a ISO 3166-1 alpha-2 (http://en.wikipedia.org/wiki/ISO_3166-1_alpha-2).

#### Default for all models

You can set the default_country_code for all models using:

```ruby
PhonyRails.default_country_code = "US"
```

### Validation

In your model use the Phony.plausible method to validate an attribute:

```ruby
validates :phone_number, phony_plausible: true
```

or the helper method:

```ruby
validates_plausible_phone :phone_number
```

this method use other validators under the hood to provide:

- presence validation using `ActiveModel::Validations::PresenceValidator`
- format validation using `ActiveModel::Validations::FormatValidator`

so we can use:

```ruby
validates_plausible_phone :phone_number, presence: true
validates_plausible_phone :phone_number, with: /\A\+\d+/
validates_plausible_phone :phone_number, without: /\A\+\d+/
validates_plausible_phone :phone_number, presence: true, with: /\A\+\d+/
```

the i18n key is `:improbable_phone`. Languages supported by default: de, en, es, fr, it, ja, kh, ko, nl, tr, ua and ru.

You can also validate if a number has the correct country number:

```ruby
validates_plausible_phone :phone_number, country_number: '61'
```

or correct country code:

```ruby
validates_plausible_phone :phone_number, country_code: 'AU'
```

You can validate against the normalized input as opposed to the raw input:

```ruby
phony_normalize :phone_number, as: :phone_number_normalized, default_country_code: 'US'
validates_plausible_phone :phone_number_normalized, presence: true, if: :phone_number?
```

Validation supports phone numbers with extension, such as `+18181231234 x1234` or `'+1 (818)151-5483 #4312'` out-of-the-box.

Return original value after validation:

The flag normalize_when_valid (disabled by default), allows to return the original phone_number when is the object is not valid. When phone validation fails, normalization is not triggered at all. It could prevent a situation where user fills in the phone number and after validation, he gets back different, already normalized phone number value, even if phone number was wrong.

Example usage:

```ruby
validates_plausible_phone :phone_number
phony_normalize :phone_number, country_code: :country_code, normalize_when_valid: true
```

Filling in the number will result with following:

When the number is incorrect (e.g. phone_number: `+44 888 888 888` for country_code 'PL'), the original validation behavior is preserved, but if the number is still invalid, the original value is returned.
When number is valid, it will save the normalized number (e.g. `+48 888 888 888` will be saved as `+48888888888`).

#### Allowing records country codes to not match phone number country codes

You may have a record specifying one country (via a `country_code` attribute) but using a phone number from another country. For example, your record may be from Japan but have a phone number from the Philippines. By default, `phony_rails` will consider your record's `country_code` as part of the validation. If that country doesn't match the country code in the phone number, validation will fail.

Additionally, `phony_normalize` will always add the records country code as the country number (eg. the user enters '+81xxx' for Japan and the records `country_code` is 'DE' then `phony_normalize` will change the number to '+4981'). You can turn this off by adding `enforce_record_country: false` to the validation options. The country_code will then only be added if no country code is specified.

If you want to allow records from one country to have phone numbers from a different one, there are a couple of options you can use: `ignore_record_country_number` and `ignore_record_country_code`. Use them like so:

```ruby
validates :phone_number, phony_plausible: { ignore_record_country_code: true, ignore_record_country_number: true }
```

Obviously, you don't have to use both, and you may not need or want to set either.

### Display / Views

In your views use:

```erb
<%= "311012341234".phony_formatted(format: :international, spaces: '-') %>
<%= "+31-10-12341234".phony_formatted(format: :international, spaces: '-') %>
<%= "+31(0)1012341234".phony_formatted(format: :international, spaces: '-') %>
```

To first normalize the String to a certain country use:

```erb
<%= "010-12341234".phony_formatted(normalize: :NL, format: :international, spaces: '-') %>
```

To return nil when a number is not valid:

```ruby
"123".phony_formatted(strict: true) # => nil
```

You can also use the bang method (phony_formatted!):

```ruby
number = "010-12341234"
number.phony_formatted!(normalize: :NL, format: :international)
number # => "+31 10 123 41234"
```

You can also easily normalize a phone number String:

```ruby
"+31 (0)30 1234 123".phony_normalized # => '31301234123'
"(0)30 1234 123".phony_normalized # => '301234123'
"(0)30 1234 123".phony_normalized(country_code: 'NL') # => '301234123'
```

Extensions are supported (identified by "ext", "ex", "x", "xt", "#", or ":") and will show at the end of the number:

```ruby
"+31 (0)30 1234 123 x999".phony_normalized # => '31301234123 x999'
"+31 (0)30 1234 123 ext999".phony_normalized # => '31301234123 x999'
"+31 (0)30 1234 123 #999".phony_normalized # => '31301234123 x999'
```

### Find by normalized number

Say you want to find a record by a phone number. Best is to normalize user input and compare to an attribute stored in the db.

```ruby
Home.find_by_normalized_phone_number(PhonyRails.normalize_number(params[:phone_number]))
```

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

Don't forget to add tests and run rspec before creating a pull request :)

See all contributors on https://github.com/joost/phony_rails/graphs/contributors.


CHANGELOG.md
# Changelog

## [v0.15.0](https://github.com/joost/phony_rails/tree/v0.15.0) (2021-04-15)

[Full Changelog](https://github.com/joost/phony_rails/compare/v0.14.13...v0.15.0)

**Closed issues:**

- normalized\_when\_valid: true seems broken if there are two phone attributes [\#195](https://github.com/joost/phony_rails/issues/195)

**Merged pull requests:**

- Test fixes - No mongoid [\#207](https://github.com/joost/phony_rails/pull/207) ([joost](https://github.com/joost))
- Updated requirements \(had to remove mongodb compatibility\) [\#206](https://github.com/joost/phony_rails/pull/206) ([joost](https://github.com/joost))
- Updated phony gem requirements to be Ruby 3 compatible [\#205](https://github.com/joost/phony_rails/pull/205) ([NikoRoberts](https://github.com/NikoRoberts))
- Make CI Green Again [\#201](https://github.com/joost/phony_rails/pull/201) ([amatsuda](https://github.com/amatsuda))

## [v0.14.13](https://github.com/joost/phony_rails/tree/v0.14.13) (2019-07-03)

[Full Changelog](https://github.com/joost/phony_rails/compare/v0.14.12...v0.14.13)

**Closed issues:**

- 0.4.11 was the latest release. However the new normalize\_when\_valid has been sitting here for a while... Release? [\#194](https://github.com/joost/phony_rails/issues/194)

**Merged pull requests:**

- Fixing \#195 - Original value should be cached in correct instance var [\#196](https://github.com/joost/phony_rails/pull/196) ([dlikhten](https://github.com/dlikhten))

## [v0.14.12](https://github.com/joost/phony_rails/tree/v0.14.12) (2019-06-21)

[Full Changelog](https://github.com/joost/phony_rails/compare/v0.14.11...v0.14.12)

**Closed issues:**

- Some German numbers not passing plausible\_numbers? without country\_code [\#193](https://github.com/joost/phony_rails/issues/193)

**Merged pull requests:**

- Upgrade Ruby versions and Relax dependancies [\#192](https://github.com/joost/phony_rails/pull/192) ([berkos](https://github.com/berkos))
- Add possibility to return original phone number when is not valid [\#190](https://github.com/joost/phony_rails/pull/190) ([synion](https://github.com/synion))
- Add UK country\_code. [\#189](https://github.com/joost/phony_rails/pull/189) ([pudiva](https://github.com/pudiva))

## [v0.14.11](https://github.com/joost/phony_rails/tree/v0.14.11) (2018-10-11)

[Full Changelog](https://github.com/joost/phony_rails/compare/v0.14.10...v0.14.11)

**Closed issues:**

- Problem with normalizing Estonian number [\#187](https://github.com/joost/phony_rails/issues/187)

**Merged pull requests:**

- Add Korean translation\(including spec\) [\#186](https://github.com/joost/phony_rails/pull/186) ([freelyageha](https://github.com/freelyageha))

## [v0.14.10](https://github.com/joost/phony_rails/tree/v0.14.10) (2018-10-11)

[Full Changelog](https://github.com/joost/phony_rails/compare/v0.14.9...v0.14.10)

**Closed issues:**

- PhonyRails.default\_country\_code too intrusive in plausible\_number? method [\#179](https://github.com/joost/phony_rails/issues/179)
- default\_country\_code overrides 00-prefix country code [\#175](https://github.com/joost/phony_rails/issues/175)
- Valid 260 area code number is reported as invalid [\#168](https://github.com/joost/phony_rails/issues/168)
- Croatian number not validating as plausible. [\#165](https://github.com/joost/phony_rails/issues/165)
- phony\_formatted not returning original String for non-digit only strings [\#163](https://github.com/joost/phony_rails/issues/163)
- should all normalized numbers be valid? [\#162](https://github.com/joost/phony_rails/issues/162)
- Extensions not working out of the box for validator [\#160](https://github.com/joost/phony_rails/issues/160)

**Merged pull requests:**

- Remove errant `puts` [\#188](https://github.com/joost/phony_rails/pull/188) ([stevenharman](https://github.com/stevenharman))
- Close issues [\#185](https://github.com/joost/phony_rails/pull/185) ([joost](https://github.com/joost))
- Better PhonyRails.plausible\_number? method. Closes \#179. [\#184](https://github.com/joost/phony_rails/pull/184) ([joost](https://github.com/joost))
- Allow numbers starting with 00 country codes. Closes \#175. [\#183](https://github.com/joost/phony_rails/pull/183) ([joost](https://github.com/joost))

## [v0.14.9](https://github.com/joost/phony_rails/tree/v0.14.9) (2018-09-05)

[Full Changelog](https://github.com/joost/phony_rails/compare/v0.14.7...v0.14.9)

**Closed issues:**

- Cut new release 0.14.7 to include new extension removal logic [\#177](https://github.com/joost/phony_rails/issues/177)
- default\_country\_code based on relation [\#174](https://github.com/joost/phony_rails/issues/174)

**Merged pull requests:**

- Allow brackets. Closes \#170. [\#182](https://github.com/joost/phony_rails/pull/182) ([joost](https://github.com/joost))
- add lambda support for default country code [\#181](https://github.com/joost/phony_rails/pull/181) ([kimyu92](https://github.com/kimyu92))
- Add Spanish translation for validation error [\#178](https://github.com/joost/phony_rails/pull/178) ([r-sierra](https://github.com/r-sierra))

## [v0.14.7](https://github.com/joost/phony_rails/tree/v0.14.7) (2018-05-25)

[Full Changelog](https://github.com/joost/phony_rails/compare/v0.14.6...v0.14.7)

**Closed issues:**

- Country code wrapped in brackets is not recognized [\#170](https://github.com/joost/phony_rails/issues/170)
- Invalid number in countryCode VN [\#159](https://github.com/joost/phony_rails/issues/159)

**Merged pull requests:**

- Extension option added to normalize\_number [\#176](https://github.com/joost/phony_rails/pull/176) ([ramaboo](https://github.com/ramaboo))
- return country from phone number [\#169](https://github.com/joost/phony_rails/pull/169) ([aovertus](https://github.com/aovertus))
- Fix code example in README to match description [\#167](https://github.com/joost/phony_rails/pull/167) ([mattruzicka](https://github.com/mattruzicka))

## [v0.14.6](https://github.com/joost/phony_rails/tree/v0.14.6) (2017-06-20)

spec/factories.rb
FactoryGirl.define do
  factory :simple_home
  factory :helpful_home
  factory :required_helpful_home
  factory :optional_helpful_home
  factory :formatted_helpful_home
  factory :not_formatted_helpful_home
  factory :australian_helpful_home
  factory :big_helpful_home
end


spec/spec_helper.rb
# frozen_string_literal: true

require 'coveralls'
Coveralls.wear!

# Own code here.

require 'rubygems'
require 'bundler/setup'

require 'active_record'
# require 'mongoid'
require 'phony_rails'

ActiveRecord::Base.establish_connection(
  adapter: 'sqlite3',
  database: ':memory:'
)

ActiveRecord::Schema.define do
  create_table :active_record_models do |table|
    table.column :phone_attribute, :string
    table.column :phone_number, :string
    table.column :phone_number_as_normalized, :string
    table.column :fax_number, :string
    table.column :country_code_attribute, :string
    table.column :symboled_phone, :string
  end
end

module SharedModelMethods
  extend ActiveSupport::Concern
  included do
    attr_accessor(
      :country_code, :country_code_attribute, :custom_country_code, :delivery_method,
      :home_country, :phone_method, :phone1_method, :recipient, :symboled_phone_method
    )

    phony_normalized_method :phone_attribute # adds normalized_phone_attribute method
    phony_normalized_method :phone_method # adds normalized_phone_method method
    phony_normalized_method :phone1_method, default_country_code: 'DE' # adds normalized_phone_method method
    phony_normalized_method :symboled_phone_method, country_code: :country_code_attribute # adds phone_with_symboled_options method
    phony_normalize :phone_number # normalized on validation
    phony_normalize :fax_number, default_country_code: 'AU'
    phony_normalize :symboled_phone, default_country_code: :country_code_attribute

    def use_phone?
      delivery_method == 'sms'
    end

    def use_email?
      delivery_method == 'email'
    end
  end
end

class ActiveRecordModel < ActiveRecord::Base
  include SharedModelMethods
end

class RelaxedActiveRecordModel < ActiveRecord::Base
  self.table_name = 'active_record_models'
  attr_accessor :phone_number, :country_code

  phony_normalize :phone_number, enforce_record_country: false
end

class ActiveRecordDummy < ActiveRecordModel
end

# In case you don't want a database for your model
class ActiveModelModel
  include ActiveModel::Model # this provides most of the interface of AR
  include ActiveModel::Validations::Callbacks # we use callbacks for normalization
  include SharedModelMethods

  # database columns don't give us free attributes, we have to define them
  attr_accessor :phone_number, :phone_attribute, :phone_number_as_normalized, :country_code_attribute, :fax_number, :symboled_phone
end

class ActiveModelDummy < ActiveModelModel
end

# class MongoidModel
#   include Mongoid::Document
#   include Mongoid::Phony
#   field :phone_attribute, type: String
#   field :phone_number,    type: String
#   field :phone_number_as_normalized, type: String
#   field :fax_number
#   field :country_code_attribute, type: String
#   field :symboled_phone, type: String
#   include SharedModelMethods
# end

# class MongoidDummy < MongoidModel
# end

I18n.config.enforce_available_locales = true

# RSpec.configure do |config|
#   # some (optional) config here
# end


lib/phony_rails.rb
# frozen_string_literal: true

require 'phony'
require 'phony_rails/string_extensions'
require 'validators/phony_validator'
require 'phony_rails/version'
require 'yaml'

module PhonyRails
  def self.default_country_code
    @default_country_code ||= nil
  end

  def self.default_country_code=(new_code)
    @default_country_code = new_code
    @default_country_number = nil # Reset default country number, will lookup next time its asked for
  end

  def self.default_country_number
    @default_country_number ||= default_country_code.present? ? country_number_for(default_country_code) : nil
  end

  def self.default_country_number=(new_number)
    @default_country_number = new_number
  end

  def self.country_number_for(country_code)
    return if country_code.nil?

    country_codes_hash.fetch(country_code.to_s.upcase, {})['country_code']
  end

  def self.country_codes_hash
    @country_codes_hash ||= YAML.load_file(File.join(File.dirname(File.expand_path(__FILE__)), 'data/country_codes.yaml'))
  end

  # This method requires a country_code attribute (eg. NL) and phone_number to be set.
  # Options:
  #   :country_number => The country dial code (eg. 31 for NL).
  #   :default_country_number => Fallback country code.
  #   :country_code => The country code we should use.
  #   :default_country_code => Some fallback code (eg. 'NL') that can be used as default (comes from phony_normalize_numbers method).
  #   :add_plus => Add a '+' in front so we know the country code is added. (default: true)
  #   :extension => Include the extension. (default: true)
  # This idea came from:
  #   http://www.redguava.com.au/2011/06/rails-convert-phone-numbers-to-international-format-for-sms/
  def self.normalize_number(number, options = {}, current_instance = nil)
    return if number.nil?

    original_number = number
    number = number.dup # Just to be sure, we don't want to change the original.
    number, ext = extract_extension(number)
    number.gsub!(/[^()\d+]/, '') # Strips weird stuff from the number
    return if number.blank?

    if _country_number = options[:country_number] || country_number_for(options[:country_code])
      options[:add_plus] = true if options[:add_plus].nil?
      # (Force) add country_number if missing
      # NOTE: do we need to force adding country code? Otherwise we can share logic with next block
      number = "#{_country_number}#{number}" if !Phony.plausible?(number) || _country_number != country_code_from_number(number)
    elsif _default_country_number = extract_default_country_number(options, current_instance)
      options[:add_plus] = true if options[:add_plus].nil?
      number = normalize_number_default_country(number, _default_country_number)
    end
    normalized_number = Phony.normalize(number)
    options[:add_plus] = true if options[:add_plus].nil? && Phony.plausible?(normalized_number)
    normalized_number = options[:add_plus] ? "+#{normalized_number}" : normalized_number

    options[:extension] = true if options[:extension].nil?
    options[:extension] ? format_extension(normalized_number, ext) : normalized_number
  rescue StandardError
    original_number # If all goes wrong .. we still return the original input.
  end

  def self.normalize_number_default_country(number, default_country_number)
    # We try to add the default country number and see if it is a
    # correct phone number. See https://github.com/joost/phony_rails/issues/87#issuecomment-89324426
    unless number =~ /\A\(?(\+|00)/ # if we don't have a + or 00
      return "#{default_country_number}#{number}" if Phony.plausible?("#{default_country_number}#{number}") || !Phony.plausible?(number) || country_code_from_number(number).nil?
      # If the number starts with ONE zero (two might indicate a country code)
      # and this is a plausible number for the default_country
      # we prefer that one.
      return "#{default_country_number}#{number.gsub(/^0/, '')}" if (number =~ /^0[^0]/) && Phony.plausible?("#{default_country_number}#{number.gsub(/^0/, '')}")
    end
    # number = "#{default_country_number}#{number}" unless Phony.plausible?(number)
    # Just return the number unchanged
    number
  end

  def self.extract_default_country_number(options = {}, current_instance = nil)
    country_code = if current_instance.present? && options[:default_country_code].respond_to?(:call)
                     options[:default_country_code].call(current_instance) rescue nil
                   else
                     options[:default_country_code]
                   end
    options[:default_country_number] || country_number_for(country_code) || default_country_number
  end

  # Returns the country dail code (eg. '31') for a number (eg. +31612341234).
  # Should probably be named 'country_number_from_number'.
  def self.country_code_from_number(number)
    return nil unless Phony.plausible?(number)

    Phony.split(Phony.normalize(number)).first
  end

  # Returns the country (eg. 'NL') for a number (eg. +31612341234).
  def self.country_from_number(number)
    return nil unless Phony.plausible?(number)

    country_codes_hash.select { |_country, hash| hash['country_code'] == country_code_from_number(number) }.keys[0]
  end

  # Wrapper for Phony.plausible?.  Takes the same options as #normalize_number.
  # NB: This method calls #normalize_number and passes _options_ directly to that method.
  # It uses the 'cc' option for Phony. This was a required param before?
  def self.plausible_number?(number, options = {})
    return false if number.blank?

    number = extract_extension(number).first
    number = normalize_number(number, options)
    country_number = options[:country_number] || country_number_for(options[:country_code]) ||
                     country_code_from_number(number) ||
                     options[:default_country_number] || country_number_for(options[:default_country_code]) ||
                     default_country_number
    Phony.plausible? number, cc: country_number
  rescue StandardError
    false
  end

  COMMON_EXTENSIONS = / *(ext|ex|x|xt|#|:)+[^0-9]*\(?([-0-9]{1,})\)?#?$/i.freeze

  def self.extract_extension(number_and_ext)
    return [nil, nil] if number_and_ext.nil?

    subbed = number_and_ext.sub(COMMON_EXTENSIONS, '')
    [subbed, Regexp.last_match(2)]
  end

  def self.format_extension(number, ext)
    ext.present? ? "#{number} x#{ext}" : number
  end

  module Extension
    extend ActiveSupport::Concern

    included do
      private

      # This methods sets the attribute to the normalized version.
      # It also adds the country_code (number), eg. 31 for NL numbers.
      def set_phony_normalized_numbers(current_instance, attributes, options = {})
        options = options.dup
        assign_values_for_phony_symbol_options(options)
        if respond_to?(:country_code)
          set_country_as = options[:enforce_record_country] ? :country_code : :default_country_code
          options[set_country_as] ||= country_code
        end
        attributes.each do |attribute|
          attribute_name = options[:as] || attribute
          raise("No attribute #{attribute_name} found on #{self.class.name} (PhonyRails)") unless self.class.attribute_method?(attribute_name)

          cache_original_attribute(current_instance, attribute) if options[:normalize_when_valid]
          new_value = PhonyRails.normalize_number(send(attribute), options, current_instance)
          current_instance.public_send("#{attribute_name}=", new_value) if new_value || attribute_name != attribute
        end
      end

      def assign_values_for_phony_symbol_options(options)
        symbol_options = %i[country_number default_country_number country_code default_country_code]
        symbol_options.each do |option|
          options[option] = send(options[option]) if options[option].is_a?(Symbol)
        end
      end
    end

    def cache_original_attribute(current_instance, attribute)
      attribute_name = "#{attribute}_original"
      current_instance.define_singleton_method("#{attribute_name}=") { |value| instance_variable_set("@#{attribute_name}", value) }
      current_instance.define_singleton_method(attribute_name) { instance_variable_get("@#{attribute_name}") }
      current_instance.public_send("#{attribute}_original=", current_instance.public_send(attribute.to_s))
    end

    module ClassMethods
      PHONY_RAILS_COLLECTION_VALID_KEYS = %i[
        add_plus
        as
        country_code
        country_number
        default_country_code
        default_country_number
        enforce_record_country
        if
        normalize_when_valid
        unless
      ].freeze
      # Use this method on the class level like:
      #   phony_normalize :phone_number, :fax_number, :default_country_code => 'NL'
      #
      # It checks your model object for a a country_code attribute (eg. 'NL') to do the normalizing so make sure
      # you've geocoded before calling this method!
      def phony_normalize(*attributes)
        options = attributes.last.is_a?(Hash) ? attributes.pop : {}
        options.assert_valid_keys(*PHONY_RAILS_COLLECTION_VALID_KEYS)
        raise ArgumentError, ':as option can not be used on phony_normalize with multiple attribute names! (PhonyRails)' if options[:as].present? && (attributes.size > 1)

        options[:enforce_record_country] = true if options[:enforce_record_country].nil?

        conditional = create_before_validation_conditional_hash(options)

        # Add before validation that saves a normalized version of the phone number
        before_validation conditional do
          set_phony_normalized_numbers(self, attributes, options)
        end
      end

      # Usage:
      #   phony_normalized_method :fax_number, :default_country_code => 'US'
      # Creates a normalized_fax_number method.
      def phony_normalized_method(*attributes)
        main_options = attributes.last.is_a?(Hash) ? attributes.pop : {}
        main_options.assert_valid_keys :country_code, :default_country_code
        attributes.each do |attribute|
          raise(StandardError, "Instance method normalized_#{attribute} already exists on #{name} (PhonyRails)") if method_defined?(:"normalized_#{attribute}")

          define_method :"normalized_#{attribute}" do |*args|
            options = main_options.merge(args.first || {})
            assign_values_for_phony_symbol_options(options)
            raise(ArgumentError, "No attribute/method #{attribute} found on #{self.class.name} (PhonyRails)") unless respond_to?(attribute)

            options[:country_code] ||= country_code if respond_to?(:country_code)
            PhonyRails.normalize_number(send(attribute), options)
          end
        end
      end

      private

      # Creates a hash representing a conditional for before_validation
      # This allows conditional normalization
      # Returns something like `{ unless: -> { attribute == 'something' } }`
      # If no if/unless options passed in, returns `{ if: -> { true } }`
      def create_before_validation_conditional_hash(options)
        if options[:if].present?
          type = :if
          source = options[:if]
        elsif options[:unless].present?
          type = :unless
          source = options[:unless]
        else
          type = :if
          source = true
        end

        conditional = {}
        conditional[type] = if source.respond_to?(:call)
                              source
                            elsif source.respond_to?(:to_sym)
                              -> { send(source.to_sym) }
                            else
                              -> { source }
                            end
        conditional
      end
    end
  end
end

# check whether it is ActiveRecord or Mongoid being used
if defined?(ActiveRecord)
  ActiveSupport.on_load(:active_record) do
    ActiveRecord::Base.send :include, PhonyRails::Extension
  end
end

ActiveModel::Model.send :include, PhonyRails::Extension if defined?(ActiveModel::Model)

# if defined?(Mongoid)
#   module Mongoid::Phony
#     extend ActiveSupport::Concern
#     include PhonyRails::Extension
#   end
# end

Dir["#{File.dirname(__FILE__)}/phony_rails/locales/*.yml"].each do |file|
  I18n.load_path << file
end


lib/phony_number.rb
require 'phony'
require "phony_number/string_extensions"
require "phony_number/version"

module PhonyNumber

  # Quick fix to get country_phone_number (phone number) for all relevant countries.
  # TODO: Replace with some gem or something.
  COUNTRY_NUMBER = {
    'NL' => '31',
    'BE' => '32',
    'DE' => '49',
    'GB' => '44',
    'FR' => '33',
    'ES' => '34',
    'IT' => '39',
    'US' => '1',
    'AU' => '61',
    'LU' => '352'
  }

  # This method requires a country_code attribute (eg. NL) and phone_number to be set.
  # Options:
  #   :country_code => The country code we should use.
  #   :default_country_code => Some fallback code (eg. 'NL') that can be used as default (comes from phony_normalize_numbers method).
  def self.normalize_number(number, options = {})
    return if number.blank?
    number = Phony.normalize(number) # TODO: Catch errors
    if country_number = COUNTRY_NUMBER[options[:country_code] || options[:default_country_code]]
      # Add country_number if missing
      number = "#{country_number}#{number}" if not number =~ /^(00|\+)?#{country_number}/
    end
    number = Phony.normalize(number)
  rescue
    number # If all goes wrong .. we still return the original input.
  end

  # This module is added to AR.
  module ActiveRecordExtension

    def self.extended(base)
      base.send :include, InstanceMethods
      base.extend ClassMethods
    end

    module InstanceMethods
  
    private

      # This methods sets the attribute to the normalized version.
      # It also adds the country_code (number), eg. 31 for NL numbers.
      def set_phony_normalized_numbers(attributes, options = {})
        options[:country_code] ||= self.country_code if self.respond_to?(:country_code)
        attributes.each do |attribute|
          write_attribute(attribute, PhonyNumber.normalize_number(read_attribute(attribute), options))
        end
      end

    end

    module ClassMethods

      # Use this method on the class level like:
      #   phony_normalize_numbers :phone_number, :fax_number, :default_country_code => 'NL'
      #
      # It checks your model object for a a country_code attribute (eg. 'NL') to do the normalizing so make sure
      # you've geocoded before calling this method!
      def phony_normalize_numbers(*attributes)
        options = attributes.last.is_a?(Hash) ? attributes.last : {} 
        attributes.each do |attribute|
          # Add before validation that saves a normalized version of the phone number
          self.before_validation do 
            set_phony_normalized_numbers(attributes, options)
          end
        end
      end

    end

  end

end
ActiveRecord::Base.extend PhonyNumber::ActiveRecordExtension

lib/phony_rails/string_extensions.rb
# frozen_string_literal: true

class String
  # Usage:
  #   "+31 (0)30 1234 123".phony_normalized # => '+31301234123'
  #   "(0)30 1234 123".phony_normalized # => '301234123'
  #   "(0)30 1234 123".phony_normalized(country_code: 'NL') # => '301234123'
  def phony_normalized(options = {})
    raise ArgumentError, "Expected options to be a Hash, got #{options.inspect}" unless options.is_a?(Hash)

    options = options.dup
    PhonyRails.normalize_number(self, options)
  end

  # Add a method to the String class so we can easily format phone numbers.
  # This enables:
  #   "31612341234".phony_formatted # => '06 12341234'
  #   "31612341234".phony_formatted(:spaces => '-') # => '06-12341234'
  # To first normalize a String use:
  #   "010-12341234".phony_formatted(:normalize => :NL)
  # To return nil when a number is not correct (checked using Phony.plausible?) use
  #   "010-12341234".phony_formatted(strict: true)
  # When an error occurs during conversion it will return the original String.
  # To raise an error use:
  #   "somestring".phone_formatted(raise: true)
  def phony_formatted(options = {})
    raise ArgumentError, "Expected options to be a Hash, got #{options.inspect}" unless options.is_a?(Hash)

    options = options.dup
    normalize_country_code = options.delete(:normalize)
    s, ext = PhonyRails.extract_extension(self)
    s = (normalize_country_code ? PhonyRails.normalize_number(s, default_country_code: normalize_country_code.to_s, add_plus: false) : s.gsub(/\D/, ''))
    return if s.blank?
    return if options[:strict] && !Phony.plausible?(s)

    PhonyRails.format_extension(Phony.format(s, options.reverse_merge(format: :national)), ext)
  rescue StandardError
    raise if options[:raise]

    s
  end

  # The bang method
  def phony_formatted!(options = {})
    raise ArgumentError, 'The :strict options is only supported in the phony_formatted (non bang) method.' if options[:strict]

    replace(phony_formatted(options))
  end
end
